pennylane/ops/qubit/state_preparation.py

```diff
@@ -497,3 +584,44 @@ def circuit():
     """int: Number of trainable parameters that the operator depends on."""
 
     grad_method = None
+
+
+def _sparse_statevec_permute_and_embed(
+    state: csr_matrix, wires: list, wire_order: list
+) -> csr_matrix:
+    """Permutes the wires of a statevector represented as a scipy.sparse.csr_matrix. If `wire_order` contains `wires`, then embed the `state` with corresponding orders, padding with bit 0 on other wires.
+
+    Args:
+        state (csr_matrix): the input statevector
+        wires (Iterable[int]): the wires of the input statevector
+        wire_order (Iterable[int]): the wires of the output statevector. E.g., [0, 2, 1] means the permutation of wires 0, 1, 2 to 0, 2, 1. wires=[2, 1] and wire_order=[1, 0, 2] means embedding the input state in a permuted order.
+
+    Returns:
+        csr_matrix: the permuted statevector
+    """
+    wires = Wires(wires)
+    wire_order = Wires(wire_order) if wire_order else wires
+
+    if not wire_order.contains_wires(wires):
+        raise WireError(
+            f"wire_order must contain all wires. Got wires {wires} and wire_order {wire_order}"
+        )
+
+    if wires == wire_order:
+        return state
+
+    index_map = _build_index_map(wires, wire_order)
+    perm_pos = index_map[state.indices]
+    new_csr = csr_matrix((state.data, perm_pos, state.indptr), shape=(1, 2 ** len(wire_order)))
+    return new_csr
+
+
+def _build_index_map(wires, wire_order):
+    n_wires = len(wires)
+    index_map = np.zeros(2**n_wires, dtype=int)
+    for pos in range(2**n_wires):
+        pos_bin = format(pos, f"0{n_wires}b")
+        wire_values_map = {wire: pos_bin[i] for i, wire in enumerate(wires)}
+        pos_bin_perm = [wire_values_map[wire] if wire in wires else "0" for wire in wire_order]
+        index_map[pos] = int("".join(pos_bin_perm), 2)
+    return index_map
```

Comment:
On previous discussion https://github.com/PennyLaneAI/pennylane/pull/6863#discussion_r1931165127
@AmintorDusko @PietropaoloFrisoni I improved the implementation a little bit. Now the workflow is: first we calc the needed `index_map`, using only the information of `wires` and `wire_order`; then we get the mapped nonzero indices; finally we use the very last instantiation interface of `csr_matrix`, `csr_matrix((data, indices, indptr), [shape=(M, N)])`, (see https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html) to get the output state.

Reply:
I don't see a direct way of gettting the permuted state via things like `return state[0, pos]` though.